File: EBook.cpp
#include "EBook.h"
#include <sstream>
#include <iomanip>

EBook::EBook(int id, const std::string& title, const std::string& author,
    double price, const std::string& isbn, const std::string& format)
    : Product(id, title, "", price, "Электронная книга"),
    author_(author), isbn_(isbn), format_(format) {
    addTag("ebook");
    addTag("digital");
}

const std::string& EBook::getAuthor() const { return author_; }
const std::string& EBook::getIsbn() const { return isbn_; }
const std::string& EBook::getFormat() const { return format_; }

std::string EBook::toDisplayString() const {
    std::ostringstream oss;
    oss << "«" << name_ << "» (" << format_ << ", " << category_ << ") — "
        << std::fixed << std::setprecision(2) << getPrice() << "₽ (электронная)";
    return oss.str();
}

double EBook::getPrice() const {
    double basePrice = Product::getPrice();
    return std::max(50.0, basePrice * 0.9);
}

std::unique_ptr<Product> EBook::clone() const {
    return std::make_unique<EBook>(*this);
}

EBook& EBook::operator=(const Product& other) {
    if (this == &other) return *this;

    id_ = other.id_;
    name_ = other.name_;
    price_ = other.price_;
    category_ = other.category_;
    description_ = other.description_;
    tags_ = other.tags_;

    return *this;
}

File: EBook.cs
using System;

namespace BookRecommender.Models
{
    /// <summary>
    /// Класс электронной книги
    /// </summary>
    public class EBook : Book, IDisposable
    {
        public string Format { get; protected set; }
        public double FileSizeMB { get; protected set; }

        private bool _disposed = false;

        public EBook() : base() { }

        public EBook(int id, string title, string author, double price, string isbn,
                    string format = "PDF", double fileSizeMB = 1.0)
            : base(id, title, author, price, isbn)
        {
            if (string.IsNullOrWhiteSpace(format))
                throw new ArgumentException("Format cannot be empty", nameof(format));
            if (fileSizeMB <= 0)
                throw new ArgumentOutOfRangeException(nameof(fileSizeMB), "File size must be positive");

            Format = format;
            FileSizeMB = fileSizeMB;
            Category = "Электронная книга";
            AddTag("ebook");
            AddTag("digital");
        }

        public EBook(string title, string author, double price, string isbn,
                    string format = "PDF", double fileSizeMB = 1.0)
            : base(title, author, price, isbn)
        {
            if (string.IsNullOrWhiteSpace(format))
                throw new ArgumentException("Format cannot be empty", nameof(format));
            if (fileSizeMB <= 0)
                throw new ArgumentOutOfRangeException(nameof(fileSizeMB), "File size must be positive");

            Format = format;
            FileSizeMB = fileSizeMB;
            Category = "Электронная книга";
            AddTag("ebook");
            AddTag("digital");
        }

        public override double GetPrice()
        {
            double basePrice = base.GetPrice();
            return Math.Max(50.0, basePrice * 0.9); // Скидка 10% на цифру
        }

        public override string ToDisplayString() =>
            $"«{Name}» ({Format}, {Category}) — {GetPrice():F2}₽ (электронная)";

        public override double ApplyDiscount(double discountPercent)
        {
            return base.ApplyDiscount(discountPercent + 10); // Доп. скидка 10%
        }

        public override object Clone() => new EBook(Id, Name, Author, Price, Isbn, Format, FileSizeMB);

        public void Dispose()
        {
            if (!_disposed)
            {
                Console.WriteLine($"EBook '{Name}' disposed.");
                _disposed = true;
            }
        }

        ~EBook() => Dispose();

        public override string ToString() => $"[EBook {Id}] {Name} — {Format} — {GetPrice():F2}₽";
    }
}